Enable `no_implicit_reexport` for core files [mypy]

cdce8p opened the pull request:
## Proposed change
Followup to https://github.com/home-assistant/core/pull/63244#issuecomment-1008017811
This PR enables the `no_implicit_reexport` config option from mypy for all core files.

https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-no-implicit-reexport

### PRs
* #63829
* #63830
* #63832


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

cdce8p committed:
Enable no_implicit_reexport for core files [mypy]

cdce8p committed:
Last changes

cdce8p commented:
@frenck I've mentioned this change earlier already. The PR enables the mypy `no_implicit_reexport` setting for all core files.

frenck commented:
How would other integrations expend/enable this?

cdce8p committed:
Merge remote-tracking branch 'upstream/dev' into mypy-no_implicit_reexport

cdce8p commented:
> How would other integrations expend/enable this?

Integrations could be added to the `NO_IMPLICIT_REEXPORT_MODULES` list in `mypy_config.py`. My hope would be that this is only temporary and at some point we enable it as default for strict typing.

https://github.com/home-assistant/core/blob/1f38dd0b5422833c6050ed5d53ae33a2f717553f/script/hassfest/mypy_config.py#L139-L143

--
Before that can happen, one thing left to decide would be how to handle imports from other packages. An example
https://github.com/home-assistant/core/blob/0fadce751f0e3cfa08e25212f0f70a9ec164b201/homeassistant/components/hdmi_cec/__init__.py#L30

Here we import `DOMAIN` from `components.media_player`. Technically it's defined in `media_player.const` and only imported in `media_player.__init__`. Since it isn't reexported nor mentioned in `__all__`, this would trigger the mypy error.
https://github.com/home-assistant/core/blob/0fadce751f0e3cfa08e25212f0f70a9ec164b201/homeassistant/components/media_player/const.py#L33

There are a few options here, some better some worse
* Change the import in `hdmi_cec.__init__` to `from ...media_player.const import DOMAIN as ...`. Not my favorite as the placement in `const` is more an implementation detail and could change at some point.
* Add an `__all__` to `media_player.__init__`. Difficult to keep in sync.?
* Explicit reexports. I.e. in `media_player.__init__` do something like `from .const import DOMAIN as DOMAIN`. I would suggest this one. Although then we're dealing with the `useless-import-alias` pylint error in most cases.
_I though about disabling it globally, but in most cases it does still make sense. It's probably better to disable it on a per-file bases (for now)._

What do you think about it?

frenck commented:
The first one is the only one that doesn't defeat the purpose of the linter rule to begin with?

cdce8p committed:
Merge remote-tracking branch 'upstream/dev' into mypy-no_implicit_reexport

cdce8p commented:
> The first one is the only one that doesn't defeat the purpose of the linter rule to begin with?

Yes. However the linter rule might be a bit too strong in this case. It just depends how you view the individual integrations within the scope of the project.

If everything should be considered internal, then yes option (1) would be the way forward. In contrast, if each integration should be considered its own little "package", then it would make sense to "hide" some implementation details like where a constant is defined. The main advantage being that we won't need to update all imports in other integrations if for example we move a constant from `__init__` to `const`.

I lean more towards the latter but this is just my personally opinion. The first would be completely valid too.

--
It would also be possible to choose exact imports and only add aliases where it's truly an improvement. That's what I've done so far if you for example take a look at the changes in `auth/permissions/__init__.py`.

https://github.com/home-assistant/core/blob/2fb80183897cd5cb88bf7a797d13a7e399a6cf6f/homeassistant/auth/permissions/__init__.py#L12-L14

The uses in `auth/models.py`
https://github.com/home-assistant/core/blob/849abaca8bbb61947bd7fa0f4923db8c466406fe/homeassistant/auth/models.py#L14

https://github.com/home-assistant/core/blob/849abaca8bbb61947bd7fa0f4923db8c466406fe/homeassistant/auth/models.py#L27

https://github.com/home-assistant/core/blob/849abaca8bbb61947bd7fa0f4923db8c466406fe/homeassistant/auth/models.py#L37

--
The PR can be merged even without us reaching a conclusion. It works with all of the options. _Although maybe we can find common ground in what I've mentioned last 🤔_

frenck commented:
I think if we don't want to import it from the actual source (e.g, the const `from ...media_player.const import DOMAIN as ...`), and considering the other cases actually defeat the purpose of the rule in the first place.

I would suggest not to further trying to achieve getting this rule enabled, as it is cleary isn't a match for us in that case.

We either forbid reexport and use the original source, or we don't. Putting rules in place to work around a linter rule, is not something we should even make standard.

cdce8p commented:
> I would suggest not to further trying to achieve getting this rule enabled, as it is cleary isn't a match for us in that case.
> 
> We either forbid reexport and use the original source, or we don't. Putting rules in place to work around a linter rule, is not something we should even make standard.

Makes sense. Although in that case, I would still recommend to enable it for the core files **only**.  The mypy check was quite helpful in detecting `from homeassistant.core impor HomeAssistantError` for example. I.e. only the PRs which have already been merged and this one, nothing more.

cdce8p committed:
Revert logic to enable no_implicit_reexport on integration files

frenck commented:
> Makes sense. Although in that case, I would still recommend to enable it for the core files only. The mypy check was quite helpful in detecting from homeassistant.core impor HomeAssistantError for example. I.e. only the PRs which have already been merged and this one, nothing more.

Don't get me wrong. I'm (personally) all in favor of enabling this linter rule. I just don't think we should work around it.

cdce8p commented:
> > Makes sense. Although in that case, I would still recommend to enable it for the core files only. The mypy check was quite helpful in detecting from homeassistant.core impor HomeAssistantError for example. I.e. only the PRs which have already been merged and this one, nothing more.
> 
> Don't get me wrong. I'm (personally) all in favor of enabling this linter rule EVERYWHERE. I just don't think we should work around it.

No offense taken. We won't be working around it though if it's limited to core files. The changes here are the last once necessary.

frenck commented:
After some thought and actual use, I would like to suggest going with the `__all__` approach.

Difficult to land maybe, but explicit. I think that explicit should always win.
First occurrences already in the code base now, I would suggest to proceed with that.

frenck commented:
Didn't mean to close it :S sorry about that.

cdce8p committed:
Merge remote-tracking branch 'upstream/dev' into mypy-no_implicit_reexport

cdce8p committed:
Revert "Revert logic to enable no_implicit_reexport on integration files"

This reverts commit 0f309e68d808b0e7ab23af4e8970cee6a9136771.

cdce8p committed:
Use __all__ for auth.permissions

cdce8p committed:
Fix new issues

cdce8p committed:
Allow no_implicit_reexport for not strictly typed modules

cdce8p commented:
> After some thought and actual use, I would like to suggest going with the `__all__` approach.
> 
> Difficult to land maybe, but explicit. I think that explicit should always win. First occurrences already in the code base now, I would suggest to proceed with that.

Works for me! I've updated the PR to re-add the ability to enable the setting for specific integration. As a first example, I though adding `homeassistant.components.diagnostics.*` might work. I've also updated `auth.permissions.__init__` to use `__all__` instead of reexports.

From my point of view, the PR would be ready and could be merged. Let me know what you think.

frenck reviewed (approved):
LGTM, Thanks, @cdce8p 👍
